Add support for body and headers  in parameters and CEL filtering

chmouel opened the pull request:
We are now able to access the full body and headers from the template variable and CEL filtering as CEL expression.

This allow to output  or filter anything GitHub or others send you in your PipelineRun

See doc for more details

<!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

# Changes

<!-- Describe your changes here- ideally you can get that description straight from
your descriptive commit message(s)! -->

# Submitter Checklist

- [ ] 📝 A good commit message is important for other reviewers to understand the context of your change. Please refer to [How to Write a Git Commit Message](https://cbea.ms/git-commit/) for more details how to write beautiful commit messages. We rather have the commit message in the PR body and the commit message instead of an external website.
- [ ] ♽ Run `make test` before submitting a PR (ie: with [pre-commit](https://pipelinesascode.com/dev/tools), no need to waste CPU cycle on CI. (or even better install [pre-commit](https://pre-commit.com/) and do `pre-commit install` in the root of this repo).
- [ ] ✨ We heavily rely on linters to get our code clean and consistent, please ensure that you have run `make lint` before submitting a PR. The [markdownlint](https://github.com/DavidAnson/markdownlint) error can get usually fixed by running `make fix-markdownlint` (make sure it's installed first)
- [ ] 📖 If you are adding a user facing feature or make a change of the behavior, please verify that you have documented it
- [ ] 🧪 100% coverage is not a target but most of the time we would rather have a unit test if you make a code change.
- [ ] 🎁 If that's something that is possible to do please ensure to check if we can add a e2e test.
- [ ] 🔎 If there is a flakiness in the CI tests then don't *necessary* ignore it, better get the flakyness fixed before merging or if that's not possible there is a good reason to bypass it. (token rate limitation may be a good reason to skip).

chmouel committed:
Add support for body and headers  in params

We are now able to access the full body and headers from the template
variable as CEL expression.

This allow to output anything GitHub or others send you in your
PipelineRun

Signed-off-by: Chmouel Boudjnah <chmouel@redhat.com>

Roming22 commented on docs/content/docs/guide/authoringprs.md:
```diff
@@ -166,14 +168,74 @@ For example if I want to have a `PipelineRun` targeting a `pull_request` but
+        #!/usr/bin/env python3
+        import json
+        labels=json.loads("""{{ body.pull_request.labels }}""")
+        for label in labels:
```
You could use a list comprehension to get rid of the for loop:
```
if 'bug' in [label['name'] for label in labels]:
```

chmouel replied:
yes! but here the  idea is to make it the most idiomatic to almost pseudo code level for the non enlightened by the tao of python :)

chmouel committed:
Add cel filtering for matching on body and headers

Further to the previous commit, this commit adds the ability to match on
body and headers from the payload as a CEL expression.

This allows to do some more advanced filtering and matching on event.

Signed-off-by: Chmouel Boudjnah <chmouel@redhat.com>

chmouel commented:
added cel filtering on body and params as well to this PR.

I have separated the two commit between cel filtering and params with CEL...  This make easier to review/changes

chmouel commented:
/retest

savitaashture commented on docs/content/docs/guide/authoringprs.md:
```diff
@@ -28,20 +29,21 @@ weight: 3
+  - `{{target_branch}}`: The branch name on which the event targets (same as `source_branch` for push events).
+  - `{{target_namespace}}`: The target namespace where the Repository has matched and the PipelineRun will be created.
+  - `{{body}}`: The full payload body available as a CEL expression (see [below](#using-the-body-and-headers-in-a-pipelines-as-code-parameter)).
+  - `{{headers}}`: The payload request headers available as a CEL expression (see [below](#using-the-body-and-headers-in-a-pipelines-as-code-parameter)).
```
nit

We sorted alphabetical order can you add `body`  and `headers` in correct place

chmouel replied:
done, but i have changed it completely as a markdown table, it does looks better but there is probably going to be an issue with width for the example, it still scrollable with the mouse if needed to see

i think we should keep as table

savitaashture commented on test/gitea_params_test.go:
```diff
@@ -185,11 +185,11 @@ func TestGiteaParamsOnRepoCR(t *testing.T) {
+func TestGiteaParamsBodyHeadersCEL(t *testing.T) {
 	topts := &tgitea.TestOpts{
 		Regexp:      successRegexp,
-		TargetEvent: "pull_request, push",
```
can we add another test for verifying `push` event for CEL expression

chmouel replied:
good idea, i reworked the whole test including plenty of comments to explain how it does it... (let's do this in the future making sure all tests scenarios are fully commented)

savitaashture commented:
@chmouel 
I have few query 
1. once we support accessing body then user can directly use  `{{ body }} only using CEL expression but not as a custom parameters https://nightly.pipelines-as-code.pages.dev/docs/guide/repositorycrd/#custom-parameter-expansion

am i right ?

2. Also with this PR changes user can directly use `{{ body }}` and access body payload inside their  pipelineruns and for that they don't need to use custom parameters is my understanding correct ?
I have tried https://github.com/savitaashture/article/pull/534/files#diff-196299b9620e4fd19a1831ae73491cc7fd2ddc307df374e7fb9cfb88cd75acfbR38 and its working as expected

savitaashture commented on docs/content/docs/guide/authoringprs.md:
```diff
@@ -28,25 +29,27 @@ weight: 3
+  - `{{source_url}}`: The source repository URL from which the event come from (same as `repo_url` for push events).
+  - `{{target_branch}}`: The branch name on which the event targets (same as `source_branch` for push events).
+  - `{{target_namespace}}`: The target namespace where the Repository has matched and the PipelineRun will be created.
+  - `{{body}}`: The full payload body available as a CEL expression (see [below](#using-the-body-and-headers-in-a-pipelines-as-code-parameter)).
```
this line indicates body available only as a cel expression

but actually apart from CEL user can directly use in Pipelineruns should we mention that as well :thinking:

chmouel replied:
done

chmouel commented:
/retest

chmouel committed:
Docs improvements and reorganization

* Convert docs variables list to markdown table
  Since it does look better and easier to sort, there is an issue with
  width on example but i think people can just select the table and it
  will be fine.

* Make the docs-dev makefile target to open the web browser when
  doing make docs-dev

* Break custom parameters in a separate page

* Reorder the menu

Signed-off-by: Chmouel Boudjnah <chmouel@redhat.com>

chmouel commented:
@savitaashture we spoke about your questions offline

the main difference between custom params and this one is being able to specify a set of static parameters from the repo cr....

custom parameters don't support body and headers as value yet... will probably submit it next to improve but didn't want to confuse further since it's already getting taking long to get this one merged

piyush-garg commented:
lgtm
